### PR TITLE
Introduce simplified builders for GraphDatabase

### PR DESCRIPTION
**Pull request by cleishm**
Includes:

| class | description |
| --- | --- |
| `org.neo4j.embedded.GraphDatabase` | An interface for a Neo4j Graph Database that holds additional methods necessary for managing the database instance itself, in addition to the graph API methods inherited from `GraphDatabaseService`. |
| `org.neo4j.embedded.CommunityGraphDatabase` | a container class for `GraphDatabase` factory methods, i.e. `#open(File)` and `#build()` (see below). |
| `org.neo4j.embedded.TestGraphDatabase` | An extension to the `GraphDatabase` interface, providing methods useful for internal testing. |
| `org.neo4j.embedded.CommunityTestGraphDatabase` | a container class for `TestGraphDatabase` factory methods, i.e. `#open(File)`, `#openEphemeral()`, `#build()` and `#buildEphemeral()`. |
| `org.neo4j.embedded.HighAvailabilityGraphDatabase` | a container class for `GraphDatabase` factory methods, that include configuration directives relevant to enterprise HA instances. |
| `org.neo4j.embedded.HighAvailabilityTestGraphDatabase` | a container class for `TestGraphDatabase` factory methods, that include configuration directives relevant to enterprise HA instances. |

The `#build()` methods return a builder that uses a fluent, programatic construction for injecting dependencies. I.e.:

```
    GraphDatabase db = CommunityGraphDatabase.build()
            .withLogProvider( customLogProvider )
            .open( new File("/var/graph.db") );
```

Notes:
- `CommunityGraphDatabase` et al are only used to hold the static factory methods - with Java 8, these methods could instead live in the interfaces themselves (e.g. `GraphDatabase`).
- Most tests have been migrated to the new factories. Some deprecated HA related factories and tests remain.
- Opens the way to leverage Rickard's recent improvements, by skipping the use of `Config` and directly adding usable objects into the Modules.

With input from @lutovich and @thobe.

### CODE REVIEW

**Comment by thobe** [comment]
Overall I like this, but I think it might be healthy to allow some time for conversation around this before it being merged.

In particular I'm unsure about whether `org.neo4j` is the appropriate package for these new classes, although I can't say I have a better alternative to offer up.

**Comment by tinwelint** [comment]
At a first glance looks to be a cleaner and more easy-to-use way of starting databases. I like the convenience methods f.ex. `allowStoreUpgrade()`

The previous `GraphDatabaseFactory` and friends live in `org.neo4j.graphdb.factory`. Something like that would probably be better than `org.neo4j`, would be my initial reaction.

File: community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+        return self();
+    }
+
+    /**
+     * Starts the graph database.
+     *
+     * @param storeDir The filesystem location for the store, which will be created if necessary
+     * @return The running database
+     */
+    public GRAPHDB open( File storeDir )
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java]: I would like overloaded `open` methods accepting `String` and potentially even `Path` type arguments. Even looking at the example in your commit message hints that that would be useful.

File: community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+    }
+
+    /**
+     * Set an internal parameter. Use of this method is not recommended.
+     *
+     * @param key parameter key
+     * @param value parameter value
+     * @return this builder
+     */
+    public BUILDER withParam( String key, String value )
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java]: I like `configuration` or `config` better than `param`. `param` is way too generic as a term and doesn't tell what it is. When I read it I had to check if that was for some internal flags or something. Externally we talk about database `configuration`, not database `params` right?

**Comment by chrisvest** [comment]
One of `CommunityTestGraphDatabase` and `TestHighAvailabilityGraphDatabase` should be renamed to be consistent with the other.

File: community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+     * The amount of memory to use for mapping the store files, in bytes. If Neo4j is running on a dedicated server,
+     * then it is generally recommended to leave about 2-4 gigabytes for the operating system, give the
+     * JVM enough heap to hold all your transaction state and query context, and then leave the rest for
+     * the page cache. If this is not set, then it is assumed that Neo4j is running exclusively on the machine,
+     * and the size will be heuristically set to 75% of RAM minus the max Java heap size.
+     *
+     * @param size the amount of memory to use for mapping the store files, in bytes
+     * @return this builder
+     */
+    public BUILDER usingPageCacheMemory( long size )
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java]: what made you use the term `using` for this configuration? Why wouldn't f.ex. `withLogicalLogRotationThreshold` be called `usingLogicalLogRotationThreshold`, or for that matter why not call this `withPageCacheMemory`?

File: community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+        return self();
+    }
+
+    /**
+     * Set the maximum number of archived logical logs.
+     *
+     * @param count the maximum number of archived logical logs
+     * @return this builder
+     */
+    public BUILDER withLogicalLogMaxArchives( int count )
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java]: This name is inconsistent with the other `withLogicalLogPruningThreshold....` methods. Consistency would be great, f.ex. `withLogicalLogPruningThresholdFileCount`

File: community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+        }
+        return self();
+    }
+
+    /**
+     * Disable the background index sampling.
+     *
+     * @return this builder
+     */
+    public BUILDER disableBackgroundIndexSampling()
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java]: To go in line with the other `with...` methods this could actually be called `withoutBackgroundIndexSampling`

**Comment by rickardoberg** [comment]
We have tons of settings, and only a few have convenience methods. Is there any predictable rule for what settings becomes methods or not?

File: community/kernel/src/main/java/org/neo4j/graphdb/GraphDatabaseService.java [hunk 6]
```diff
@@ -27,24 +27,12 @@
...
- * <code>GraphDatabaseService graphDb = new GraphDatabaseFactory().newEmbeddedDatabase( "var/graphDb" );
- * // ... use Neo4j
- * graphDb.{@link #shutdown() shutdown()};</code>
- * </pre>
- * <p/>
- * GraphDatabaseService provides operations to {@link #createNode() create
- * nodes}, {@link #getNodeById(long) get nodes given an id} and ultimately {@link #shutdown()
- * shutdown Neo4j}.
+ * The main access point for working with a Graph Database. The most common
+ * implementation is the {@link org.neo4j.GraphDatabase} class, which is used to
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/graphdb/GraphDatabaseService.java]: You refer to `org.neo4j.GraphDatabase` as if it was a class.

File: community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java (hunk 1 above)
Comment (cleishm) [inline community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java]: I'd prefer not to have `String` - it's not the responsibility of this library to figure out how to parse a string into a File / Path. I totally agree on having `Path`, but I'd want the code to respect the filesystem attached to the `Path` object, and not throw it away by converting to `File` internally: so I thought it might have to wait until you and @chrisvest have finished changing everything to use `Path` internally?

File: community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java (hunk 2 above)
Comment (cleishm) [inline community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java]: I'm happy to call it `withConfig`. However, I'm also thinking it should be as obtuse as possible - since it really shouldn't be used by the end-user, and is only exposed as a transitional thing (indeed, now I'm wondering if we could/should only expose it in the `TestGraphDatabase`?).

File: community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java (hunk 3 above)
Comment (cleishm) [inline community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java]: Any of those alternatives are fine. The names I picked were arbitrary and not overly thought through (yet).

File: community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java (hunk 4 above)
Comment (cleishm) [inline community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java]: +1. If you'd like to propose some alternatives for them all, let's update it?

File: community/kernel/src/main/java/org/neo4j/graphdb/GraphDatabaseService.java (hunk 6 above)
Comment (cleishm) [inline community/kernel/src/main/java/org/neo4j/graphdb/GraphDatabaseService.java]: Good catch

**Comment by cleishm** [comment]
@tinwelint: _"...friends live in org.neo4j.graphdb.factory. Something like that would probably be better than org.neo4j..."_ -- Unsurprisingly, I prefer just `org.neo4j`. From an end user perspective, this is a more obvious place. It is the Neo4j Graph Database. :)
@chrisvest: _"One of CommunityTestGraphDatabase and TestHighAvailabilityGraphDatabase should be renamed to be consistent with the other."_ -- +1
@rickardoberg: _"We have tons of settings, and only a few have convenience methods. Is there any predictable rule for what settings becomes methods or not?"_ -- Yes. All of them :) Either in `GraphDatabaseBuilder`, if they're supposed to be visible to the end user, or `TestGraphDatabaseBuilder`, if they're really only for internal use. That way we can, eventually, replace the use of `Config` with the direct injection of values into your Platform & Edition modules.

File: community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java (hunk 1 above)
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java]: Fair points, perhaps it's not a good idea.

**Comment by rickardoberg** [comment]
@cleishm The problem with "all of them" is that this is undefined, as kernel extensions can also have settings. But as long as there is a way to pass in any string param, that's ok. We need to retain Config for this reason, at least.

File: community/kernel/src/main/java/org/neo4j/GraphDatabase.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j;
+
+import org.neo4j.graphdb.GraphDatabaseService;
+import org.neo4j.graphdb.event.KernelEventHandler;
+import org.neo4j.graphdb.event.TransactionEventHandler;
+
+/**
+ * A running Neo4j graph database.
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/GraphDatabase.java]: I'd like a better documentation about the difference between `GraphDatabaseService` and `GraphDatabase`

File: community/kernel/src/test/java/org/neo4j/test/TestGraphDatabaseRule.java [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+import org.neo4j.graphdb.QueryExecutionException;
+import org.neo4j.graphdb.Result;
+import org.neo4j.graphdb.Transaction;
+import org.neo4j.graphdb.mockfs.EphemeralFileSystemAbstraction;
+import org.neo4j.graphdb.schema.Schema;
+import org.neo4j.helpers.Pair;
+import org.neo4j.io.fs.FileUtils;
+import org.neo4j.logging.LogProvider;
+
+public class TestGraphDatabaseRule extends ExternalResource implements Supplier<TestGraphDatabase>
```
Comment (tinwelint) [inline community/kernel/src/test/java/org/neo4j/test/TestGraphDatabaseRule.java:45]: Wouldn't it be neat if `TestGraphDatabaseRule` implemented `GraphDatabase` directly so that tests wouldn't have to get the db from it and tuck it away in a variable? Otherwise the test ends up having two db variables, where one is named `...Rule`. I'd love to get away from that. Of course that could be another PR after this one, but it's just that this PR already changes all these test classes, so why not?

File: community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java (hunk 2 above)
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java]: I see your point there. Perhaps it's alright.

File: community/kernel/src/test/java/org/neo4j/test/TestGraphDatabaseRule.java (hunk 8 above)
Comment (cleishm) [inline community/kernel/src/test/java/org/neo4j/test/TestGraphDatabaseRule.java:45]: Yes - the same thought did cross my mind. I'd love to do that in a separate PR?

File: community/kernel/src/main/java/org/neo4j/GraphDatabase.java (hunk 7 above)
Comment (cleishm) [inline community/kernel/src/main/java/org/neo4j/GraphDatabase.java]: +1

**Comment by jakewins** [comment]
I'm not sure I'm +1 on this change, it seems like making users rewrite a lot of code for no functional benefit other than naming and package locations. But I seem to be in the minority.. Some comments though:
- I don't think we should use the root `org.neo4j` namespace. This is not the end-all API, it's factories for constructing embedded database instances. I'd suggest we keep it under `org.neo4j.graphdb`, where the rest of the embedded API is.
- There's a lot of work that has gone into Config to move away from java-centric mechanisms and to have a strong configuration service that works on the DBMS level, who decided that we are removing that? Config does a lot more than just pass parameters around, I quite like it.
- -1 on having all configuration options hard-coded in the interfaces. It makes modular configuration hard like Rickard notes, and it makes these interfaces huge, which makes it hard to write alternative implementations.

**Comment by lutovich** [comment]
Imho this change is quite nice. It introduces a much more pleasant way to create embedded databases than current 'factory that accepts builder' does.
I'm a bit scared by the idea to put all config convenience methods in interfaces. This will make those interfaces really huge. It feels like having `withParam(String, String)` is valuable and only most valuable config options should get to the interface.

**Comment by tinwelint** [comment]
I'm also +1 on only having a small set of most common config methods on the builders, like `allowStoreUpgrade` and memory settings.

**Comment by cleishm** [comment]
It's an interesting thought. My take is slightly different - the configuration exists, whether it's made directly via a method or in a string parameter to a `withParam` call, and thus it is part of the defined interface. Having them as methods over strings gives benefits of type safety, clarity and discoverability (via Javadoc and via IDE hinting). And whilst having them all as methods certainly does mean many methods making that interface huge, that's only exposing the truth that there are many configuration options - options that any implementation must be prepared to support. Having fewer options as methods and another means for setting the rest is only hiding that reality.

But I'm happy to change it, if nobody else feels the same way :)

**Comment by lutovich** [comment]
Hm, actually having all config options as methods might not be that bad. Wide interface is generally bad but this are builders and it feels unrealistic that we'll need to write alternative implementations, mocks or anything like that.

**Comment by jakewins** [comment]
If these builders were the primary or only way to construct database instances, I'd probably agree. However, the main usage path for configuration is when running neo4j proper, passing configuration in config files. If we used hard-coded methods on these interfaces, we'd have two config surfaces to manually maintain. We'd need a third mechanism to pass all the internal config options, and we'd expose a large set of types to users that we currently do not, making the embedded config inconsistent with the string syntax for general configuration that users will find in the manual.

Also -1 on having "some" config this way and "some" config via the config system. I think that increases confusion (where should I look for config setting x?), rather than helping users.

Eg; if we were building an embedded library only, this makes sense. But this is a secondary mechanism in a larger product. I think that the benefits of method-completion is outweighed by the confusion of config for embedded not being the same as Neo4j configuration. The current scheme provides IDE completion without these drawbacks.

**Comment by cleishm** [comment]
At some point the string parameters need to be parsed and converted into typed values that the kernel/embedded uses. My thought would be to do that in the server only, and let the kernel/embedded use only the typed values (basically, it's dependency injection). That's helpful for injecting pre-configured things also, like entire extensions, shared page caches, shared cluster communication clients, etc.

File: community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+        return self();
+    }
+
+    /**
+     * Set the default cypher parser version.
+     *
+     * @param version The version number for the cypher parser version to use by default
+     * @return this builder
+     */
+    public BUILDER usingCypherParserVersion( String version )
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java]: Still uses `using`. I'd like the consistency of `with`, better to be careful here since methods introduced here will be user facing. Also a couple of other methods below.

File: community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+        return self();
+    }
+
+    /**
+     * Set the maximum number of archived internal logs to retain.
+     *
+     * @param count the maximum number of archived internal logs to retain
+     * @return this builder
+     */
+    public BUILDER withInternalLogMaxArchives( int count )
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java]: I wouldn't think this deserves a place as a method. There have been discussions about which config params should be visible, and that only the most common should be. I think we should start with a small set and grow slowly and carefully.

Comment (cleishm) [inline community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java]: There's been a couple of suggestions that only the most common options should be methods. I'm not sure I buy that though, because then the rest have to be set via string interpretation -- which doesn't really make sense in a programatic API, unless you're trying to hide those as internal-only options. It does make sense in server, where values are read from a properties file. To save on argument though, let's merge this with a minimal set while the discussion continues (I can move the additional ones to a new PR). Is it just this method you're questioning?

Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java]: I think there are more, but the final set of methods could be decided after this PR goes in. Remove which ever "unimportant" params and we'll reiterate afterwards.

**Comment by cleishm** [comment]
(I accidentally pushed an empty branch instead of a rebased one, so github helpfully closed this PR. Reopened.)

**Comment by davidegrohmann** [comment]
@cleishm this needs rebase

**Comment by rickardoberg** [comment]
Closed for now, as it has been decided that now is not the right time to introduce another way to instantiate embedded databases.
